### PR TITLE
Macro expansion protocol implementation example

### PR DESCRIPTION
**Pull request by jakemac53**
Followup to https://github.com/dart-lang/language/pull/2021. This partially implements one possible version of the MacroExecutor interface, and is a port of my early experiments using `IsolateMirror.loadUri` to load and execute macros.

I could fully flesh this out, if the implementation teams think that would be helpful.

### CODE REVIEW

File: working/macros/api/src/protocol/isolate_mirror_impl.dart [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    return GenericResponse(error: e);
+  }
+}
+
+Future<GenericResponse<MacroExecutionResult>> _executeDefinitionsPhase(
+    ExecuteDefinitionsPhaseRequest request) async {
+  try {
+    var instance = _macroInstances[request.macro];
+    if (instance == null) {
+      print(_macroInstances);
```
Comment (johnniwinther) [inline working/macros/api/src/protocol/isolate_mirror_impl.dart:84]: Remove this - but maybe add [_macroInstances] to the exception message.

File: working/macros/api/src/protocol/isolate_mirror_impl.dart [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+  operator ==(other) => other is _MacroInstanceIdentifier && id == other.id;
+
+  int get hashCode => id;
+}
+
+/// Our implementation of [MacroExecutionResult].
+class _MacroExecutionResult implements MacroExecutionResult {
+  @override
+  final List<DeclarationCode> agumentations = <DeclarationCode>[];
```
Comment (johnniwinther) [inline working/macros/api/src/protocol/isolate_mirror_impl.dart:133]: augmentations

File: working/macros/api/src/protocol/isolate_mirror_impl.dart [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+          ]),
+          ', ',
+        ],
+        ']',
+      ],
+      if (declaration.namedParameters.isNotEmpty) ...[
+        '{',
+        for (var named in declaration.namedParameters) ...[
+          ParameterCode.fromParts([
+            named.type.code,
```
Comment (johnniwinther) [inline working/macros/api/src/protocol/isolate_mirror_impl.dart:197]: You need support for 'required'.

File: working/macros/test/isolate_mirror_executor/simple_macro.dart [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  SimpleMacro.named({this.x, this.y});
+
+  @override
+  FutureOr<void> buildDefinitionForFunction(
+      FunctionDeclaration method, FunctionDefinitionBuilder builder) {
+    if (method.namedParameters
+        .followedBy(method.positionalParameters)
+        .isNotEmpty) {
+      throw ArgumentError(
+          'This macro can only be ran on functions with no arguments!');
```
Comment (johnniwinther) [inline working/macros/test/isolate_mirror_executor/simple_macro.dart:25]: 'ran' -> 'run'

**Review by johnniwinther — APPROVED** [review]

File: working/macros/api/src/protocol/isolate_mirror_executor.dart [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  @override
+  Future<MacroExecutionResult> executeDefinitionsPhase(
+          MacroInstanceIdentifier macro,
+          Declaration declaration,
+          TypeComparator typeComparator,
+          ClassIntrospector classIntrospector,
+          TypeIntrospector typeIntrospector) =>
+      _sendRequest(ExecuteDefinitionsPhaseRequest(macro, declaration,
```
Comment (scheglov) [inline working/macros/api/src/protocol/isolate_mirror_executor.dart:99]: We cannot send `TypeComparator` for example to another isolate, can we? It potentially a large data structure that references a big portion of analyzer or CFE. We probably have to marshal requests back and forth.

File: working/macros/api/src/protocol/isolate_mirror_impl.dart [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+Future<GenericResponse<MacroClassIdentifier>> _loadMacro(
+    LoadMacroRequest request) async {
+  try {
+    var identifier = _MacroClassIdentifier(request.library, request.name);
+    if (_macroClasses.containsKey(identifier)) {
+      throw UnsupportedError(
+          'Reloading macros is not supported by this implementation');
+    }
+    var libMirror =
+        await currentMirrorSystem().isolate.loadUri(request.library);
```
Comment (scheglov) [inline working/macros/api/src/protocol/isolate_mirror_impl.dart:45]: Can we load a macro from string in memory?

File: working/macros/api/src/protocol/isolate_mirror_executor.dart (hunk 5 above)
Comment (jakemac53) [reply working/macros/api/src/protocol/isolate_mirror_executor.dart:99]: Correct - a more real implementation would I think have its own internal rpcs to handle this type of introspection - which would delegate on the `executor` end of things to these objects. In fact in my actual original prototype that is exactly what it does.

This is just a starting point, more illustrative of how things are expected to work together than anything else at the moment.

File: working/macros/api/src/protocol/isolate_mirror_impl.dart (hunk 6 above)
Comment (jakemac53) [reply working/macros/api/src/protocol/isolate_mirror_impl.dart:45]: I suppose possibly through a `data:` uri? But I haven't tried. This api today actually doesn't even work for `package:` uris although that is just a bug.

File: working/macros/api/src/protocol/isolate_mirror_impl.dart [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+              named.defaultValue!,
+            ],
+          ]),
+          ', ',
+        ],
+        '}',
+      ],
+      ') ',
+      body,
+    ]));
```
Comment (munificent) [inline working/macros/api/src/protocol/isolate_mirror_impl.dart:213]: A++++ use of control flow elements. :D

**Review by munificent — APPROVED** [review]

File: working/macros/api/src/protocol/isolate_mirror_impl.dart (hunk 7 above)
Comment (jakemac53) [reply working/macros/api/src/protocol/isolate_mirror_impl.dart:213]: haha ya this one got a bit crazy but was able to make it work lol! A bit tricky not being able to introduce new variables, but if we can get [binding expressions](https://github.com/dart-lang/language/issues/1210) that could maybe make these even more useful!

File: working/macros/api/src/protocol/isolate_mirror_impl.dart [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+          positionalRequired.type.code,
+          ' ',
+          positionalRequired.name,
+        ]),
+        ', '
+      ],
+      if (declaration.positionalParameters.any((p) => !p.isRequired)) ...[
+        '[',
+        for (var positionalOptional
+            in declaration.positionalParameters.where((p) => p.isRequired)) ...[
```
Comment (jakemac53) [inline working/macros/api/src/protocol/isolate_mirror_impl.dart:184]: Note to self lol:

```suggestion
            in declaration.positionalParameters.where((p) => !p.isRequired)) ...[
```

**Review by scheglov — APPROVED** [review]

**Pull request merged**